teps/0137-cloudevents-controller.md

```diff
@@ -0,0 +1,527 @@
+separate controller. The new controller will support new versions and types
+of events:
+
+- Tekton v2 events, which embed `v1.TaskRun` and `v1.PipelineRun` as opposed
+  to v1 events which embed the corresponding `v1beta1` resources
+- [CDEvents core events][cdevents-core]
+- User driven [CDEvents][cdevents] of any kind
+
+## Motivation
+
+When Tekton introduced support for events, the [CloudEvents][cloudevents]
+format was chosen as it provides an interoperable standard for the format of
+the enveloper of events. At the time, no such standard existed for the
+payload. Because of that I created, together with other people on the
+[CDF Interoperability Special Interest Group][cdf-sig-interop], the
+[CDEvents][cdevents] project. The TEP proposes to adopt that standard, now that
+it is available, to enhance Tekton interoperability with other tools in the
+CI/CD space.
+
+Rather then introducing CDEvents support in the core Tekton controllers, this
+TEP additionally proposes to split the event feature into a dedicated controller,
+work which has already started for the `CustomRun` events, and which provides a
+better separation of concerns and security benefits, as it separates the secrets
+which are needed by `TaskRun` and `PipelineRun` controllers from those required
+to successfully deliver events to sinks.
+
+Finally, the new controller enables the development of new features in the area
+of events and notification with no or reduced impact on the `TaskRun` and `PipelineRun`
+controller, which means the code in those controller remains leaner, and their
+scalability is not compromised.
+
+### Goals
+
+- Produce Tekton v2 events through the dedicate event controller
+- Produce core CDEvents through the dedicate event controller
+- Give task authors a way to trigger additional events from their tasks. Tekton
+  has no knowledge about abstractions like builds, artifacts, tests and more,
+  that are modelled by CDEvents. Tekton Task authors write task that implement
+  such abstraction. Goal of this TEP is to let such authors notify about them
+  through CDEvents.
+- Give users the ability to select through configuration which event type they
+  would like to send, if any
+- Give Tekton operators the ability to remove event support from their
+  Tekton deployment
+
+### Non-Goals
+
+- Additional messaging features that may be implemented in future, like multiple
+  sinks, different event bindings, event signing and more
+- Changes to the operator to support disabling the events deployment are welcome,
+  but not part of this TEP
+
+### Use Cases
+
+- Notifications: Events produced by Tekton can be used to send notifications via
+  different channels (GitHub, Slack, etc)
+- Interoperability: CDEvents produced by Tekton can be consumed by other tools to
+  trigger the next step in the workflow. Similar events may be triggered by other
+  orchestrators like Jenkins and Argo Workflows, letting user switch between tools or
```

Comment:
Where are the other orchestrators in process of adopting CDEvents? Have they implemented them?

Reply:
Some more info about CDEvents.

A list of implementations:
- Jenkins can produce CDEvents
- ArgoCD has an [RFC](https://github.com/argoproj/argo-cd/pull/13723), I plan on collaborating on one for Argo workflows too
- Spinnaker proposal has been accepted, and the implementation is under review
- Testkube has been implemented
- Tekton has CDEvents events in the events controller in the experimental repo, this TEP proposes moving that to official support

The CDEvents project is also working with other communities, like Harbor (CNCF container registry), the proposal is close to being merged and I demoed the implementation already. We are also collaborating with CNCF, OpenSSF and VSMI.

CDEvents benefits from contribution from 20 different companies, including IBM, Ericsson, Apple, Fidelity, SAS and many more.